### PR TITLE
Adds optimization to reorder basic blocks, fixes #50

### LINKED ISSUE #50: Better flattening of basic blocks
```c
kernel void if1 (global float a[]) {
  int id = get_global_id(0);
  if ((id & 0x01) == 0)
    a[id] = 3.0;
}
```

```asm
// Module with 1 kernels, global data with 0 words (64-bit each), starting at offset 1 words and 0 words of stack-frame
// Kernel 'if1' with 46 instructions, offset 2, with following parameters: __global out float* a (4 B, 1 items) (lSize, lids, gidX, offX)
// label %start_of_function
or r1, unif, unif
or r2, unif, unif
or r0, unif, unif
or ra2, unif, unif
or ra1, unif, unif
// label %tmp.0
or ra0, r0, r0; v8min r0, r1, r1
ldi r1, 255
and r3, r0, r1
or r2, ra2, ra2; v8min r0, r2, r2
and r1, r0, r1
mul24 r0, ra0, r3
add r0, r2, r0
add r1, r0, r1
nop.never 
or r0, r1, r1
and.setf -, r0, 1 (1)
xor.ifzc r0, 1 (1), 1 (1); v8min.ifz r0, 1 (1), 1 (1)
or.setf -, elem_num, r0
brr.ifallzc (pc+4) + 4 // to %tmp.1      --- (1)
nop.never 
nop.never 
nop.never 
brr.ifanyz (pc+4) + 10 // to %tmp.2    --- (2)
nop.never 
nop.never 
nop.never 
// label %tmp.1
or r0, r1, r1
shl r0, r0, 2 (2)
or -, mutex_acq, mutex_acq
ldi vpw_setup, vpm_setup(size: 16 words, stride: 1 rows, address: h32(0))
itof vpm, 3 (3)
ldi vpw_setup, vdw_setup(rows: 1, elements: 1 words, address: h32(0))
ldi vpw_setup, vdw_setup(stride: 0)
add vpw_addr, ra1, r0
or -, vpw_wait, vpw_wait
or mutex_rel, 1 (1), 1 (1)
// label %end_of_function
or r0, unif, unif
or.setf -, elem_num, r0
brr.ifallzc (pc+4) + -42 // to %start_of_function
nop.never 
nop.never 
nop.never 
not irq, qpu_num
nop.thrend.never 
nop.never 
nop.never 
```

In this code, the branch of (1) can be removed. We should,

- check condition of (2) first
- if the condition is false, fall through to the basic block (of writing to memory)

Anyway, we need better flattening of basic blocks.
Currently, we just use the iterator of traverseing all instructions.
We need more carefully flattening, which is keen to the shape of the graph.

### PR DESCRIPTION
**Pull request by doe300**

### CODE REVIEW

**Comment by nomaddo** [comment]
Can you describe what is changed?
I compiled `50_block_flattening.cl` and compare the outputs compiled by one of `master` and `reorder_blocks`, but there is no different.

**Comment by doe300** [comment]
Really? I have to check that.
The optimization reorders the basic block within a kernel in a way, that blocks following eachother (in the control flow) are adjacent. This allows for more branches to be replaced with implicit fall-throughs and therefore also for more basic blocks to be merged.

File: testing/bugs/50_block_flattening.cl [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// See https://github.com/doe300/VC4C/issues/50
+kernel void if1 (global float a[]) {
+  int id = get_global_id(0);
+  if ((id & 0x01) == 0)
+    a[id] = 3.0;
+}
+
+void kernel test(float global * a, const int n){
+  float sum = 0.0;
+  for (int i = 0; i < 10; i++)
```
Comment (nomaddo) [inline testing/bugs/50_block_flattening.cl]: I think this `10` should be `n` (my mistake?)

**Comment by doe300** [comment]
You are right, for this particular test there is no difference. But for e.g. `test_branches.cl` you will see a few instructions saved.

**Comment by nomaddo** [comment]
Thanks, your changes seems good to me. 
In my understanding, the patch resolves limited scope of the alien order of basic-blocks: reorder a basic-block and its predesossor if predecessor is unique.

It seems okey to merge it, but we need to continue to work on the problem......

**Pull request merged**